File: ios/Reguerta/Reguerta/Presentation/Access/AuthErrorMapping.swift
import Foundation

enum AuthErrorFlow: Sendable {
    case signIn
    case signUp
    case passwordReset
}

struct AuthErrorPresentation: Sendable {
    let emailErrorKey: String?
    let passwordErrorKey: String?
    let globalMessageKey: String?

    init(
        emailErrorKey: String? = nil,
        passwordErrorKey: String? = nil,
        globalMessageKey: String? = nil
    ) {
        self.emailErrorKey = emailErrorKey
        self.passwordErrorKey = passwordErrorKey
        self.globalMessageKey = globalMessageKey
    }
}

func mapAuthFailure(_ reason: AuthSignInFailureReason, flow: AuthErrorFlow) -> AuthErrorPresentation {
    switch flow {
    case .signIn:
        switch reason {
        case .invalidEmail:
            AuthErrorPresentation(emailErrorKey: AccessL10nKey.feedbackEmailInvalid)
        case .invalidCredentials:
            AuthErrorPresentation(passwordErrorKey: AccessL10nKey.authErrorInvalidCredentials)
        case .emailAlreadyInUse:
            AuthErrorPresentation(emailErrorKey: AccessL10nKey.authErrorEmailAlreadyInUse)
        case .weakPassword:
            AuthErrorPresentation(passwordErrorKey: AccessL10nKey.authErrorWeakPassword)
        case .userNotFound:
            AuthErrorPresentation(emailErrorKey: AccessL10nKey.authErrorUserNotFound)
        case .userDisabled:
            AuthErrorPresentation(emailErrorKey: AccessL10nKey.authErrorUserDisabled)
        case .tooManyRequests:
            AuthErrorPresentation(globalMessageKey: AccessL10nKey.authErrorTooManyRequests)
        case .network:
            AuthErrorPresentation(globalMessageKey: AccessL10nKey.authErrorNetwork)
        case .unknown:
            AuthErrorPresentation(globalMessageKey: AccessL10nKey.authErrorUnknown)
        }

    case .signUp:
        switch reason {
        case .invalidEmail:
            AuthErrorPresentation(emailErrorKey: AccessL10nKey.feedbackEmailInvalid)
        case .invalidCredentials:
            AuthErrorPresentation(passwordErrorKey: AccessL10nKey.authErrorInvalidCredentials)
        case .emailAlreadyInUse:
            AuthErrorPresentation(emailErrorKey: AccessL10nKey.authErrorEmailAlreadyInUse)
        case .weakPassword:
            AuthErrorPresentation(passwordErrorKey: AccessL10nKey.authErrorWeakPassword)
        case .userNotFound:
            AuthErrorPresentation(emailErrorKey: AccessL10nKey.authErrorUserNotFound)
        case .userDisabled:
            AuthErrorPresentation(emailErrorKey: AccessL10nKey.authErrorUserDisabled)
        case .tooManyRequests:
            AuthErrorPresentation(globalMessageKey: AccessL10nKey.authErrorTooManyRequests)
        case .network:
            AuthErrorPresentation(globalMessageKey: AccessL10nKey.authErrorNetwork)
        case .unknown:
            AuthErrorPresentation(globalMessageKey: AccessL10nKey.authErrorUnknown)
        }

    case .passwordReset:
        switch reason {
        case .invalidEmail:
            AuthErrorPresentation(emailErrorKey: AccessL10nKey.feedbackEmailInvalid)
        case .userNotFound:
            AuthErrorPresentation(emailErrorKey: AccessL10nKey.authErrorUserNotFound)
        case .userDisabled:
            AuthErrorPresentation(emailErrorKey: AccessL10nKey.authErrorUserDisabled)
        case .tooManyRequests:
            AuthErrorPresentation(globalMessageKey: AccessL10nKey.authErrorTooManyRequests)
        case .network:
            AuthErrorPresentation(globalMessageKey: AccessL10nKey.authErrorNetwork)
        case .unknown, .invalidCredentials, .emailAlreadyInUse, .weakPassword:
            AuthErrorPresentation(globalMessageKey: AccessL10nKey.authErrorUnknown)
        }
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/presentation/access/AuthErrorMapping.kt
package com.reguerta.user.presentation.access

import androidx.annotation.StringRes
import com.reguerta.user.R
import com.reguerta.user.domain.access.AuthSignInFailureReason

enum class AuthErrorFlow {
    SIGN_IN,
    SIGN_UP,
    PASSWORD_RESET,
}

data class AuthErrorUiMapping(
    @param:StringRes val emailErrorRes: Int? = null,
    @param:StringRes val passwordErrorRes: Int? = null,
    @param:StringRes val globalMessageRes: Int? = null,
)

fun mapAuthFailure(reason: AuthSignInFailureReason, flow: AuthErrorFlow): AuthErrorUiMapping =
    when (flow) {
        AuthErrorFlow.SIGN_IN -> when (reason) {
            AuthSignInFailureReason.INVALID_EMAIL ->
                AuthErrorUiMapping(emailErrorRes = R.string.feedback_email_invalid)
            AuthSignInFailureReason.INVALID_CREDENTIALS ->
                AuthErrorUiMapping(passwordErrorRes = R.string.auth_error_invalid_credentials)
            AuthSignInFailureReason.EMAIL_ALREADY_IN_USE ->
                AuthErrorUiMapping(emailErrorRes = R.string.auth_error_email_already_in_use)
            AuthSignInFailureReason.WEAK_PASSWORD ->
                AuthErrorUiMapping(passwordErrorRes = R.string.auth_error_weak_password)
            AuthSignInFailureReason.USER_NOT_FOUND ->
                AuthErrorUiMapping(emailErrorRes = R.string.auth_error_user_not_found)
            AuthSignInFailureReason.USER_DISABLED ->
                AuthErrorUiMapping(emailErrorRes = R.string.auth_error_user_disabled)
            AuthSignInFailureReason.TOO_MANY_REQUESTS ->
                AuthErrorUiMapping(globalMessageRes = R.string.auth_error_too_many_requests)
            AuthSignInFailureReason.NETWORK ->
                AuthErrorUiMapping(globalMessageRes = R.string.auth_error_network)
            AuthSignInFailureReason.UNKNOWN ->
                AuthErrorUiMapping(globalMessageRes = R.string.auth_error_unknown)
        }

        AuthErrorFlow.SIGN_UP -> when (reason) {
            AuthSignInFailureReason.INVALID_EMAIL ->
                AuthErrorUiMapping(emailErrorRes = R.string.feedback_email_invalid)
            AuthSignInFailureReason.INVALID_CREDENTIALS ->
                AuthErrorUiMapping(passwordErrorRes = R.string.auth_error_invalid_credentials)
            AuthSignInFailureReason.EMAIL_ALREADY_IN_USE ->
                AuthErrorUiMapping(emailErrorRes = R.string.auth_error_email_already_in_use)
            AuthSignInFailureReason.WEAK_PASSWORD ->
                AuthErrorUiMapping(passwordErrorRes = R.string.auth_error_weak_password)
            AuthSignInFailureReason.USER_NOT_FOUND ->
                AuthErrorUiMapping(emailErrorRes = R.string.auth_error_user_not_found)
            AuthSignInFailureReason.USER_DISABLED ->
                AuthErrorUiMapping(emailErrorRes = R.string.auth_error_user_disabled)
            AuthSignInFailureReason.TOO_MANY_REQUESTS ->
                AuthErrorUiMapping(globalMessageRes = R.string.auth_error_too_many_requests)
            AuthSignInFailureReason.NETWORK ->
                AuthErrorUiMapping(globalMessageRes = R.string.auth_error_network)
            AuthSignInFailureReason.UNKNOWN ->
                AuthErrorUiMapping(globalMessageRes = R.string.auth_error_unknown)
        }

        AuthErrorFlow.PASSWORD_RESET -> when (reason) {
            AuthSignInFailureReason.INVALID_EMAIL ->
                AuthErrorUiMapping(emailErrorRes = R.string.feedback_email_invalid)
            AuthSignInFailureReason.USER_NOT_FOUND ->
                AuthErrorUiMapping(emailErrorRes = R.string.auth_error_user_not_found)
            AuthSignInFailureReason.USER_DISABLED ->
                AuthErrorUiMapping(emailErrorRes = R.string.auth_error_user_disabled)
            AuthSignInFailureReason.TOO_MANY_REQUESTS ->
                AuthErrorUiMapping(globalMessageRes = R.string.auth_error_too_many_requests)
            AuthSignInFailureReason.NETWORK ->
                AuthErrorUiMapping(globalMessageRes = R.string.auth_error_network)
            AuthSignInFailureReason.UNKNOWN,
            AuthSignInFailureReason.INVALID_CREDENTIALS,
            AuthSignInFailureReason.EMAIL_ALREADY_IN_USE,
            AuthSignInFailureReason.WEAK_PASSWORD,
            ->
                AuthErrorUiMapping(globalMessageRes = R.string.auth_error_unknown)
        }
    }
